### PR TITLE
add local inference enabling/disabling setting

### PR DESCRIPTION
**Pull request by rbhavna**
### Description
add a setting to enable/disable local model inference
 
### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/2164
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLPredictionAction.java [hunk 1]
```diff
@@ -121,6 +122,8 @@ public RestChannelConsumer prepareRequest(RestRequest request, NodeClient client
     MLPredictionTaskRequest getRequest(String modelId, String algorithm, RestRequest request) throws IOException {
         if (FunctionName.REMOTE.name().equals(algorithm) && !mlFeatureEnabledSetting.isRemoteInferenceEnabled()) {
             throw new IllegalStateException(REMOTE_INFERENCE_DISABLED_ERR_MSG);
+        } else if (FunctionName.isDLModel(FunctionName.from(algorithm)) && !mlFeatureEnabledSetting.isLocalModelInferenceEnabled()) {
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/rest/RestMLPredictionAction.java:125]: Can we also add check in transport action?

File: plugin/src/main/java/org/opensearch/ml/settings/MLFeatureEnabledSetting.java [hunk 2]
```diff
@@ -46,4 +53,12 @@ public boolean isAgentFrameworkEnabled() {
         return isAgentFrameworkEnabled;
     }
 
+    /**
+     * Whether the local model feature is enabled. If disabled, APIs in ml-commons will block local model inference.
+     * @return whether the agent framework is enabled.
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/settings/MLFeatureEnabledSetting.java:58]: `agent framework`?

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLPredictionAction.java (hunk 1 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/rest/RestMLPredictionAction.java:125]: yes makes sense. Will do

File: plugin/src/main/java/org/opensearch/ml/settings/MLFeatureEnabledSetting.java (hunk 2 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/settings/MLFeatureEnabledSetting.java:58]: Sorry, will change it

**Comment by dhrubo-os** [comment]
The field name`local_model_inference_enabled` doesn't seem quite right to me because it suggests that only prediction is being managed. In reality, we are also controlling model registration and deployment.

Therefore, names like local_model_management_enabled, local_model_lifecycle_enabled, or local_model_services_enabled seem more appropriate for the operations we want to handle.

However, I realize these suggestions might not align perfectly with the existing remote_model_inference_enabled field.

File: plugin/src/main/java/org/opensearch/ml/action/deploy/TransportDeployModelAction.java [hunk 3]
```diff
@@ -143,6 +144,8 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<MLDepl
                 Boolean isHidden = mlModel.getIsHidden();
                 if (functionName == FunctionName.REMOTE && !mlFeatureEnabledSetting.isRemoteInferenceEnabled()) {
                     throw new IllegalStateException(REMOTE_INFERENCE_DISABLED_ERR_MSG);
+                } else if (FunctionName.isDLModel(functionName) && !mlFeatureEnabledSetting.isLocalModelInferenceEnabled()) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/deploy/TransportDeployModelAction.java:147]: Don't we need to add this check for transport registration and transport prediction?

Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/action/deploy/TransportDeployModelAction.java:147]: yes working on it now

File: plugin/src/main/java/org/opensearch/ml/action/deploy/TransportDeployModelAction.java [hunk 4]
```diff
@@ -144,7 +144,7 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<MLDepl
                 Boolean isHidden = mlModel.getIsHidden();
                 if (functionName == FunctionName.REMOTE && !mlFeatureEnabledSetting.isRemoteInferenceEnabled()) {
                     throw new IllegalStateException(REMOTE_INFERENCE_DISABLED_ERR_MSG);
-                } else if (FunctionName.isDLModel(functionName) && !mlFeatureEnabledSetting.isLocalModelInferenceEnabled()) {
+                } else if (FunctionName.isDLModel(functionName) && !mlFeatureEnabledSetting.isLocalModelEnabled()) {
                     throw new IllegalStateException(LOCAL_MODEL_DISABLED_ERR_MSG);
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/deploy/TransportDeployModelAction.java:148]: just wondering if we should throw `IllegalStateException` or Forbidden exception?

File: plugin/src/main/java/org/opensearch/ml/utils/MLExceptionUtils.java [hunk 5]
```diff
@@ -20,6 +20,8 @@ public class MLExceptionUtils {
     public static final String NOT_SERIALIZABLE_EXCEPTION_WRAPPER = "NotSerializableExceptionWrapper: ";
     public static final String REMOTE_INFERENCE_DISABLED_ERR_MSG =
         "Remote Inference is currently disabled. To enable it, update the setting \"plugins.ml_commons.remote_inference_enabled\" to true.";
+    public static final String LOCAL_MODEL_DISABLED_ERR_MSG =
+        "Local Model Inference is currently disabled. To enable it, update the setting \"plugins.ml_commons.local_model.enabled\" to true.";
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/utils/MLExceptionUtils.java:24]: Should we also change this part? As we are not just disable inference

File: plugin/src/main/java/org/opensearch/ml/action/deploy/TransportDeployModelAction.java (hunk 4 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/action/deploy/TransportDeployModelAction.java:148]: We can maybe use the same one as remote inference and agent framework to stay consistent?

File: plugin/src/main/java/org/opensearch/ml/utils/MLExceptionUtils.java (hunk 5 above)
Comment (rbhavna) [reply plugin/src/main/java/org/opensearch/ml/utils/MLExceptionUtils.java]: fixed it

**Review by ylwu-amzn — APPROVED** [review]

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**